Rebase of #4772 (url link tool by CruzeBlade)

rolandlo opened the pull request:
This is a rebase of @CruzeBlade's url link tool from #4772 on current master. I have squashed all commits into one so to simplify the rebase process.
I will continue to make it ready for merging if that's what @CruzeBlade wants me to do.
It would be a pitty if this great feature was bit-rotting for longer.

@CruzeBlade Please tell me if you are fine with that. As I understood you are lacking time to finish work on this feature.

CruzeBlade committed:
Url Link Tool (rebase of xournalpp#4772)

rolandlo committed:
Make link dialog GTK4-compatible

rolandlo committed:
Fixup some warnings

rolandlo committed:
Replace gtk_message_dialog by XojMsgBox

rolandlo committed:
Fix page deletion

rolandlo committed:
Avoid highlighting already highlighted link

rolandlo committed:
Fix border computation

rolandlo committed:
Fix clipboard pasting

rolandlo committed:
Fix link selection after closing Link Editor

rolandlo committed:
Add InsertUndoAction for links

rolandlo committed:
Allow editing link from double clicking on selection

rolandlo committed:
Fix link editor termination

rolandlo committed:
Add link tool to most toolbars

rolandlo committed:
Activate link tool via menubar (or CTRL+SHIFT+u)

Photon89 commented:
I just tested the Linux AppImage, works great for me!

searumblinguminari commented:
I've only tested the Windows package X64 x86_64.
It seems to work as intended.
However, I don't understand the correct behavior of "Change the orientation (left, center, right) of a link." It seems that no matter which I select, there is no change.

rolandlo committed:
Fix typo

rolandlo committed:
Split off LinkPopover class from LinkEditor class

rolandlo committed:
Put LinkPopover class into src/core/gui

rolandlo committed:
Rename LinkEditor to LinkHandler

rolandlo committed:
Bump file format version

rolandlo commented:
I have bumped the file format version (from 4 to 5), since documents containing links will be saved without links when saved in an older Xournal++ version.

A link element looks like this in a .xopp-file
```xml
<link align="0" font="Sans" size="12" x="117" y="116" color="#ff00ffff" url="https%3A%2F%2Fsbb.ch" text="Fahrplan"/>
```

rolandlo commented:
> In the long run it would be a nice to have radio buttons
> 
> [ ] Use same color as text color [ ] Use different color: (color chooser here)
> 
> But the new tool works very well as is already, so the idea shouldn't delay merging the PR!

Thanks for the suggestion. I agree. This will have to wait for another PR. The color info is already contained in the file format, so we don't run into troubles when allowing setting the color later.

rolandlo committed:
Simplify layoutToggled

rolandlo committed:
Use Builder class in LinkDialog

rolandlo committed:
Fix mem leak in LinkDialog::getFont

rolandlo committed:
Simplify LinkDialog::isUrlValid and LinkDialog::isTextValid

rolandlo commented:
The code is now cleaned up and functional enough that I feel it is ready to be reviewed/merged after some squashing of commits.

There is one bug with popover arrows that I could not solve. Sometimes when switching between selection of two links I get popover arrows like this: 

<img width="681" height="352" alt="linkpopover" src="https://github.com/user-attachments/assets/bf977462-5b9a-4cf1-bb17-545c3e3abe5a" />

It might very well be a Gtk bug. There are some bug reports on the Gtk issue tracker related to popover arrows, but I'm not sure which one of these might be relevant. I think it's not a dealbraker in any case.

rolandlo committed:
Make placeholders consistent

bhennion commented on src/core/control/tools/LinkHandler.cpp:
```diff
@@ -0,0 +1,183 @@
+            page->fireElementChanged(link);
+            GdkWindow* window = gtk_widget_get_window(view->getWidget());
+            GdkCursor* cursor = gdk_cursor_new_from_name(gdk_window_get_display(window), "hand2");
+            gdk_window_set_cursor(window, cursor);
```
This should probably be done via our XournalppCursor

bhennion commented on src/core/control/xojfile/LoadHandler.cpp:
```diff
@@ -732,6 +733,37 @@ void LoadHandler::parseTexImage() {
+
+    this->link->setAlignment(align);
+
+    this->link->setText(std::string(g_uri_unescape_string(text, NULL)));
```
The text could have non-ASCII characters and is UTF-8-encoded, so this should be made into a std::u8string

rolandlo replied:
Excuse my ignorance. Before I try to make text and url into std::u8strings. What can actually go wrong here? I have tested with non ASCII characters on Windows and haven't run into a problem, even when saving and loading. We don't iterate over the characters or do stuff like that with those strings. Do you have a way how to get a failure?

bhennion commented on src/core/gui/LinkPopover.cpp:
```diff
@@ -0,0 +1,61 @@
+
+LinkPopover::~LinkPopover() {
+    gtk_widget_destroy(GTK_WIDGET(this->label));
+    gtk_widget_destroy(GTK_WIDGET(this->popover));
```
This is not GTK4-compatible.

Also, the label is owned by the box which is owned by the popover, so destroying the popover will destroy the label already. No need to do both.

bhennion commented on src/core/model/Link.cpp:
```diff
@@ -0,0 +1,148 @@
+    link->snappedBounds = this->snappedBounds;
+    link->sizeCalculated = this->sizeCalculated;
+    link->highlighted = this->highlighted;
+    return std::move(link);
```
```suggestion
    return link;
```

bhennion commented on src/core/gui/dialog/LinkDialog.h:
```diff
@@ -0,0 +1,92 @@
+public:
+    void preset(XojFont font, std::string text, std::string url, LinkAlignment layout = LinkAlignment::LEFT);
+    std::string getText();
+    std::string getURL();
```
Same concern here about UTF-8. Those should be u8string

bhennion reviewed (changes requested):
Here is a partial review. I did not test it and I only read parts of the files.

rolandlo committed:
Unify URL activation

rolandlo commented on src/core/gui/dialog/LinkDialog.cpp:
```diff
@@ -0,0 +1,184 @@
+
+    Builder builder(control->getGladeSearchPath(), UI_FILE);
+
+    this->linkDialog = static_cast<xoj::util::raii::GtkWindowUPtr>(GTK_WINDOW(builder.get(UI_DIALOG_NAME)));
```
Done!

rolandlo commented on src/core/gui/inputdevices/PenInputHandler.cpp:
```diff
@@ -519,6 +519,8 @@ void PenInputHandler::actionLeaveWindow(InputEvent const& event) {
         if (!this->inputContext->getSettings()->getInputSystemDrawOutsideWindowEnabled()) {
             this->actionEnd(this->lastHitEvent);
         }
+    } else if (toolHandler->getToolType() == TOOL_LINK) {
```
Also without the special handling after closing the link editor it's not enough to press into the document again to create a new link. One has to click twice.

rolandlo committed:
Fixup link selection

Co-authored-by: Benjamin Hennion <benjamin.hennion@wanadoo.fr>

rolandlo committed:
Fix link unhighlighting

bhennion commented on src/core/control/tools/LinkHandler.cpp:
```diff
@@ -0,0 +1,171 @@
+                    const auto layer = page->getSelectedLayer();
+                    layer->addElement(std::move(linkOwn));
+                    doc->unlock();
+                    page->firePageChanged();
```
firePageChanged is overkill here

bhennion commented on src/core/gui/inputdevices/StylusInputHandler.cpp:
```diff
@@ -86,7 +86,6 @@ auto StylusInputHandler::handleImpl(InputEvent const& event) -> bool {
         cursor->updateCursor();
     }
 
-
```
Could you revert this isolated change?

rolandlo committed:
Make LinkPopover Gtk 4 compatible

rolandlo committed:
Add group undo action when changing link

Co-authored-by: Benjamin Hennion <benjamin.hennion@wanadoo.fr>

rolandlo committed:
Use OverlayView for displaying rectangular boxes around highlighted and selected links

bhennion commented on src/core/model/Element.cpp:
```diff
@@ -111,6 +111,12 @@ auto Element::distanceTo(double x, double y) const -> double {
+auto Element::containsPoint(double x, double y) const -> bool {
+    bool horizontal = x >= getX() && x <= getX() + getElementWidth();
+    bool vertical = y >= getY() && y <= getY() + getElementHeight();
+    return horizontal && vertical;
```
Because of lazy boolean evaluation, the (aknowledgely less readable) one-liner would be faster here:

```suggestion
    return x >= getX() && x <= getX() + getElementWidth() && y >= getY() && y <= getY() + getElementHeight();
```
Even better would be to ensure that sizeCalculated == true only once.

rolandlo committed:
Simplify startEditing, select and highlight methods

Also remove unnecessary testing for x < 0 || y < 0

rolandlo commented:
> You'll also need to rebase on latest once #6670 has been merged

Done! I was a unsure about what to do with the link alignment enum that needed to be made compatible with the LoadHandler. I put it into Link.h, since it is used in the Link class and LinkDialog class mostly.

rolandlo commented on src/core/model/Element.h:
```diff
@@ -72,6 +72,7 @@ class Element: public Serializable {
     virtual bool intersectsArea(double x, double y, double width, double height) const;
     /// Returns the distance between the element "as drawn" and the point (x,y)
     virtual double distanceTo(double x, double y) const;
+    virtual bool containsPoint(double x, double y) const;
```
I renamed it to `hasBoundingBoxContaining`, haven't used it in other places though and would like to defer that to another PR.

rolandlo commented on src/core/gui/PageView.cpp:
```diff
@@ -669,6 +706,12 @@ auto XojPageView::onButtonReleaseEvent(const PositionInputData& pos) -> bool {
+    } else if (control->getToolHandler()->getToolType() == TOOL_LINK) {
+        if (x < 0 || y < 0) {
+            return false;
+        }
```
I didn't find a clear reason, so I removed it.

rolandlo commented on src/core/gui/PageView.cpp:
```diff
@@ -648,6 +666,25 @@ void XojPageView::deleteView(xoj::view::OverlayView* view) {
+        this->startEditingOnButtonRelease = false;
+        if (x < 0 || y < 0) {
+            return false;
+        }
```
I didn't find a clear reason, so I removed it.

rolandlo commented on src/core/control/tools/LinkHandler.cpp:
```diff
@@ -0,0 +1,221 @@
+            }
+            this->highlightPopover->linkTo(nullptr);
+        }
+    }
```
I have simplified all of `startEditing`, `highlight` and `select`. It should be more readable now and still show the same behavior.

rolandlo commented on src/core/control/tools/LinkHandler.cpp:
```diff
@@ -0,0 +1,221 @@
+                    layer->addElement(std::move(linkOwn));
+                    auto [orig, elementIndex] = layer->removeElement(linkElement);
+                    doc->unlock();
+                    page->fireElementChanged(linkElement);  // rerender region around previous element
```
Done!

rolandlo commented on src/core/control/tools/LinkHandler.cpp:
```diff
@@ -0,0 +1,221 @@
+                    link->setFont(dlg->getFont());
+                    link->setX(linkElement->getX());
+                    link->setY(linkElement->getY());
+                    page->fireElementChanged(link);
```
Done! See below.

rolandlo commented:
One thing I'm unsure about is the way that links are saved in the underlying xml of the xopp-file. Currently it's:

```xml
<text font="Sans" size="18" x="123" y="123" color="#000000ff">Hello Text</text>
```
for text versus
```xml
<link align="0" font="Sans" size="12" x="123" y="123" color="#ff00ffff" url="https%3A%2F%2Fxournalpp.github.io" text="Hello%20Link"/>
```
for link.

So we have a text attribute for link element but not so for a text element.

Would it be better to store this as
```xml
<link align="0" font="Sans" size="12" x="123" y="123" color="#ff00ffff" url="https%3A%2F%2Fxournalpp.github.io">Hello Link</link>
```

rolandlo commented:
Closes #4732

bhennion commented on src/core/control/tools/LinkHandler.cpp:
```diff
@@ -0,0 +1,203 @@
+}
+
+void LinkHandler::startEditing(const PageRef& page, const int x, const int y) {
+    Link* linkElement = findLinkAtPos(page, x, y);
```
We should lock the document mutex for this (as a shared_lock)

bhennion commented on src/core/control/tools/LinkHandler.cpp:
```diff
@@ -0,0 +1,203 @@
+                    link->setAlignment(dlg->getLayout());
+                    link->setFont(dlg->getFont());
+                    link->setX(linkElement->getX());
+                    link->setY(linkElement->getY());
```
I'm not sure about this, but better safe than sorry I guess:
When this callback is called, the linkElement may have been destroyed. I think we should record the original position before the dialog pops up (and while the doc mutex is locked) and never dereference linkElement in the callback.

bhennion commented on src/core/control/tools/LinkHandler.cpp:
```diff
@@ -0,0 +1,203 @@
+                    doc->lock();
+                    const auto layer = page->getSelectedLayer();
+                    layer->addElement(std::move(linkOwn));
+                    auto [orig, elementIndex] = layer->removeElement(linkElement);
```
Note that this use of linkElement is fine: the pointer acts as an identifier and is never dereferenced.

bhennion commented on src/core/control/tools/LinkHandler.cpp:
```diff
@@ -0,0 +1,203 @@
+}
+
+void LinkHandler::select(const PageRef& page, const int x, const int y, const bool controlDown, XojPageView* pageView) {
+    Link* link = findLinkAtPos(page, x, y);
```
Same here: lock the mutex

bhennion commented on src/core/control/tools/LinkHandler.cpp:
```diff
@@ -0,0 +1,203 @@
+}
+
+void LinkHandler::highlight(const PageRef& page, const int x, const int y, XojPageView* pageView) {
+    Link* link = findLinkAtPos(page, x, y);
```
and here

bhennion commented on src/core/gui/dialog/LinkDialog.h:
```diff
@@ -0,0 +1,94 @@
+
+    static constexpr char PLACEHOLDER_HTTPS[] = "xournalpp.github.io";
+    static constexpr char PLACEHOLDER_MAIL[] = "email-address@provider.domain";
+    static constexpr char PLACEHOLDER_FILE[] = "/absolute/path/to/file";
```
Should this be a typical Windows path on Windows?

rolandlo replied:
Even on Windows it must start with a slash. Afterwards it doesn't matter if slashes or backslashes are used, but something like 
```term
/C:\Users\Joe\Documents\pic.jpg
``` 
looks strange, even though it works just like 
```term
/C:/Users/Joe/Documents/pic.jpg
```

Not sure how to communicate that to Windows users.

bhennion replied:
> Even on Windows it must start with a slash

Why is that?

rolandlo replied:
The general format for file URI is `file://host/path`, 
> where host is the fully qualified domain name of the system on which the path is accessible, and path is a hierarchical directory path of the form directory/directory/.../name. If host is omitted, it is taken to be "localhost", the machine from which the URL is being interpreted.

See https://en.wikipedia.org/wiki/File_URI_scheme.

bhennion replied:
I see. Then I guess something like `/C:\foo\bar` would be better already.
We may also prepend the `/` ourselves (like adding it to the prefix, on Windows).

rolandlo replied:
I thought about prepending `/` on Windows, too, but that excludes network locations like
```term
file://server/folder/data.xml
```
for those who are aware of it. I guess
```term
/C:/foo/bar
```
is then better, since it's just a typical example and doesn't look as strange as mixing slashes and backslashes.

bhennion reviewed (commented):
Here are a couple of comments.

bhennion commented:
> One thing I'm unsure about is the way that links are saved in the underlying xml of the xopp-file. Currently it's:
> 
> ```
> <text font="Sans" size="18" x="123" y="123" color="#000000ff">Hello Text</text>
> ```
> 
> for text versus
> 
> ```
> <link align="0" font="Sans" size="12" x="123" y="123" color="#ff00ffff" url="https%3A%2F%2Fxournalpp.github.io" text="Hello%20Link"/>
> ```
> 
> for link.
> 
> So we have a text attribute for link element but not so for a text element.
> 
> Would it be better to store this as
> 
> ```
> <link align="0" font="Sans" size="12" x="123" y="123" color="#ff00ffff" url="https%3A%2F%2Fxournalpp.github.io">Hello Link</link>
> ```

Do both support the same situations? Typically, are `\n`'s supported in the current link version?

It would be good to add those to the test cases. Including cases with utf8 characters in the displayed name and corner cases like that.

rolandlo commented:
> > One thing I'm unsure about is the way that links are saved in the underlying xml of the xopp-file. Currently it's:
> > ```
> > <text font="Sans" size="18" x="123" y="123" color="#000000ff">Hello Text</text>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > for text versus
> > ```
> > <link align="0" font="Sans" size="12" x="123" y="123" color="#ff00ffff" url="https%3A%2F%2Fxournalpp.github.io" text="Hello%20Link"/>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > for link.
> > So we have a text attribute for link element but not so for a text element.
> > Would it be better to store this as
> > ```
> > <link align="0" font="Sans" size="12" x="123" y="123" color="#ff00ffff" url="https%3A%2F%2Fxournalpp.github.io">Hello Link</link>
> > ```
> 
> Do both support the same situations? Typically, are `\n`'s supported in the current link version?
> 
> It would be good to add those to the test cases. Including cases with utf8 characters in the displayed name and corner cases like that.

I think that both would support the same situations. Newlines are definitely supported in the current link versions (there would be no point in aligning links to the left, center or right otherwise), but they are stored as:
```xml
<link ... text="Multiline%0ALink"/>
```
so the newline character is escaped, which wouldn't be the case with the other form.

rolandlo committed:
Lock document mutex while finding link and make file placeholder more specific on Windows

rolandlo committed:
Fixup save handler. Avoid escaping/unescaping uri and text.

bhennion commented:
> I think that both would support the same situations.

Then both are good, I guess.

bhennion commented:
It all looks good to me, with the following reserve: it would be better to extend the load/save and ser/des test units to also test for this new element type.

rolandlo committed:
Fixup serializing/deserializing link alignment

rolandlo committed:
Add unit tests for loading and serializing/deserializing

rolandlo committed:
Make link text the content of a link tag rather than an attribute

rolandlo commented:
> It all looks good to me, with the following reserve: it would be better to extend the load/save and ser/des test units to also test for this new element type.

Done! I also fixed that the link alignment wasn't serialized/deserialized.

rolandlo commented:
> > I think that both would support the same situations.
> 
> Then both are good, I guess.

I have switched to using the link text as content rather than an attribute. The main reason is consistency: for all 4 element types we had there was a content in the xml and the content was what was really displayed (text type: the text, image type: the image, teximage type: the image of the tex formula, stroke: the stroke points). Another (although less important) reason is that for attributes there is more escaping needed, which makes it less readable.

I have also removed the use of `g_uri_escape_string` and `g_uri_unescape_string` for saving/loading. There is no reason to escape text and url just for saving them in the file.

rolandlo commented:
@bhennion Please check the last 4 commits. I'm now settled with the way links are represented in the xml. This is crucial, since it would be hard to change later on.

bhennion commented on test/unit_tests/control/LoadHandlerTest.cpp:
```diff
@@ -786,3 +787,32 @@ TEST(ControlLoadHandler, testRelativePath) {
+        auto* link = dynamic_cast<const Link*>(elements[i]);
+        ASSERT_NE(link, nullptr);
+        EXPECT_STREQ(link->getText().c_str(), char_cast(text));
+        EXPECT_STREQ(link->getUrl().c_str(), char_cast(url));
```
Shouldn't alignment be check here as well?

rolandlo replied:
Yes, indeed. I have added that.

bhennion reviewed (commented):
One small thing. Other than that, LGTM!
I agree that having the link as content feels more appropriate (without any hard argument though).

rolandlo committed:
Test for link alignment in load handler test

rolandlo commented:
Merging in 48 hours if no objections are raised (and the checks are successful).
Thanks a lot for all the review @bhennion

rolandlo commented:
The URL link tool is finally merged. Thanks @CruzeBlade for all the initial work on the feature.
